### PR TITLE
tree-wide: add git revisions to alpine git urls

### LINKED ISSUE #220258: Build failure: musl
### Steps To Reproduce

1. Override the `storeDir` and `stateDir` using an overlay, [example](https://yann.hodique.info/blog/nix-in-custom-location).
2. Install Nix twice:
2.a `nix-env -i nix nss-cacert` (building Nix that targets new store, but lives in old store)
2.b `nix-env -i nix nss-cacert` (building Nix that targets and lives in new store)
3. Wait for error.

### Build log

```
[...]
error: hash mismatch in fixed-output derivation '/[redacted]/nix/store/dxbjii0l26rymrg02wgnpa7f8zr9zqyg-sys-cdefs.h.drv':
         specified: sha256-MLttfg4LYfzZXYMMN2yCmmFLzkaDwbl+BsIB7Cxug5o=
            got:    sha256-FbHdhnEKTF/AZU29hxllsidHTrf1RTHuuv/+jBBh8Q4=
[...]
```

### Additional context

I'm trying to build Nix with a different store path, so it's trying to build musl, but it fails due to a hash mismatch.
Current hash `16l3dqnfq0f20rzbkhc38v74nqcsh9n3f343bpczqq8b1rz6vfrh` in https://github.com/NixOS/nixpkgs/blob/c90c4025bb6e0c4eaf438128a3b2640314b1c58d/pkgs/os-specific/linux/musl/default.nix#L6-L9 should actually be `03pic488rzpzpbp32igmnx74f9xjclcqggadcp05yk0af63dvc8m` since that file was recently [updated](https://git.alpinelinux.org/aports/commit/main/libc-dev/sys-cdefs.h?id=30f3fb2b217d711c7343a78b185b035a2418b3b5). However, I was told it would be better to include the commit ref in the URL instead of relying on the main branch.

### Notify maintainers

@thoughtpolice @dtzWill

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.11-76060111-generic, Pop!_OS, 22.04 LTS, nobuild`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - channels(user): `"nixpkgs"`
 - nixpkgs: `/home/user/.nix-defexpr/channels/nixpkgs`

```

### PR DESCRIPTION
**Pull request by yu-re-ka**
###### Description of changes

#220258 

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Artturin** [comment]
@ofborg build musl

**Comment by Artturin** [comment]
the following now works
```
sudo mkdir -p /test-nix/var/nix && sudo chown --recursive $USER:$USER /test-nix

$(nix build --impure --expr 'with import <nixpkgs> { }; (nix.override { storeDir = "/test-nix/store"; stateDir = "/test-nix/var"; }).out' --no-link --print-out-paths)/bin/nix build ".#musl"
```

File: pkgs/development/libraries/presage/default.nix [hunk 1]
```diff
@@ -26,7 +26,7 @@ stdenv.mkDerivation rec {
 
   patches = [
     (fetchpatch {
-      url = "https://git.alpinelinux.org/aports/plain/community/presage/gcc6.patch";
+      url = "https://git.alpinelinux.org/aports/plain/community/presage/gcc6.patch?id=40e2044c9ecb36eacb3a1fd043f09548d210dc01";
```
Comment (dotlambda) [inline pkgs/development/libraries/presage/default.nix:29]: To make it zero rebuilds:
```suggestion
      name = "gcc6.patch";
      url = "https://git.alpinelinux.org/aports/plain/community/presage/gcc6.patch?id=40e2044c9ecb36eacb3a1fd043f09548d210dc01";
```

**Review by dotlambda — COMMENTED** [review]
That's too many rebuilds for master. Please revert.

**Comment by Artturin** [comment]
> That's too many rebuilds for master. Please revert.

100-500 is ok for master

**Comment by dotlambda** [comment]
> > That's too many rebuilds for master. Please revert.
> 
> 100-500 is ok for master

Oh, I didn't see that th force push lowered the number.
Still, zero would have been much preferred.

**Comment by dotlambda** [comment]
I'm adding a `name` in https://github.com/NixOS/nixpkgs/pull/220729.

**Pull request merged**